README.md
# Eligible

[![Circle CI](https://circleci.com/gh/eligible/eligible-ruby.svg?style=svg)](https://circleci.com/gh/eligible/eligible-ruby) [![Code Climate](https://codeclimate.com/github/eligible/eligible-ruby/badges/gpa.svg)](https://codeclimate.com/github/eligible/eligible-ruby)

Ruby bindings for the [Eligible API](https://eligible.com/rest)

## Installation

Add this line to your application's Gemfile:

    gem 'eligible'

And then execute:

    $ bundle

Or install it yourself with:

    $ gem install eligible

## Usage

### Setup
```ruby
require 'eligible'
Eligible.api_key = 'YOUR_KEY'
```

### Test
```ruby
Eligible.test = true
```

### Parameters overwrite

On each api call, you can overwrite the api key or the test parameter:

```ruby
Eligible::Coverage.get({:api_key => 'NEW_KEY', :test => false})
```

### Response Format

By default, all responses are in JSON, but you can request raw
access to X12 by adding is as a parameter on the api call:

```ruby
Eligible::Coverage.get({:format => "x12"})
```

# Important notes

## Payer List for Eligibility

the parameter `payer_id`, required for most of the api calls, is
provided by Eligible from its website, in xml and json format, which
you can embed into your applications.

[https://eligible.com/resources/payers/eligibility.xml](https://eligible.com/resources/payers/eligibility.xml)

[https://eligible.com/resources/payers/eligibility.json](https://eligible.com/resources/payers/eligibility.json)

## Payer List for Claims

the parameter `payer_id`, required for claims, is provided by Eligible
from its website, in xml and json format, which you can embed into
your applications.

## Medical

[https://eligible.com/resources/payers/claims/medical.xml](https://eligible.com/resources/payers/claims/medical.xml)

[https://eligible.com/resources/payers/claims/medical.json](https://eligible.com/resources/payers/claims/medical.json)

## Institutional

[https://eligible.com/resources/payers/claims/institutional.xml](https://eligible.com/resources/payers/claims/institutional.xml)

[https://eligible.com/resources/payers/claims/institutional.json](https://eligible.com/resources/payers/claims/institutional.json)

## Dental

[https://eligible.com/resources/payers/claims/dental.xml](https://eligible.com/resources/payers/claims/dental.xml)

[https://eligible.com/resources/payers/claims/dental.json](https://eligible.com/resources/payers/claims/dental.json)

## Service Type Codes

the parameter `service_type`, required on the api calls, is provided
by Eligible from its website, in xml and json format, which you can
embed into your applications.

[https://eligible.com/resources/service-codes.xml](https://eligible.com/resources/service-codes.xml)  
[https://eligible.com/resources/service-codes.json](ttps://eligible.com/resources/service-codes.json)

## Place of Service

[https://eligible.com/resources/place_of_service.json](https://eligible.com/resources/place_of_service.json)

## Health Care Provider Taxonomy

[https://eligible.com/resources/health-care-provider-taxonomy-code-set.json](https://eligible.com/resources/health-care-provider-taxonomy-code-set.json)

### Api Call Results

On all results you can check for errors in `result.error`. The raw
json format is available by using `result.to_hash`.

```ruby
coverage = Eligible::Coverage.get(params)
coverage.error
coverage.to_hash
```

## Coverage

### Reference

[https://reference.eligible.com/#coverage](https://reference.eligible.com/#coverage)

### Retrieve eligibility and benefit information

```ruby
params = {
  :service_type => "33",
  :network => "OUT",
  :payer_id   => "000001",
  :provider_last_name => "Last",
  :provider_first_name => "First",
  :provider_npi => "12345678",
  :member_id => "12345678",
  :member_last_name => "Austen",
  :member_first_name => "Jane",
  :member_dob => "1955-12-14"
}

coverage = Eligible::Coverage.get(params)
coverage.to_hash # returns all coverage info for the request
coverage.error   # return error, if any
```

## Cost Estimate

### Reference
[https://eligible.com/reference/cost-estimates](https://eligible.com/reference/cost-estimates)

### Retrieve cost estimate information

```ruby
params = {
  service_type: '98',
  network: 'IN',
  payer_id: '00001',
  provider_npi: '1234567893',
  member_id: 'COST_ESTIMATE_001',
  member_dob: '1886-01-01',
  provider_price: '200',
  level: 'individual'
}

cost_estimate = Eligible::Coverage.cost_estimate(params)
cost_estimate.to_hash # returns all coverage info along with cost estimate
```

## Medicare

### Reference

[https://reference.eligible.com/#medicare](https://reference.eligible.com/#medicare)

### Retrieve eligibility and benefit information from CMS Medicare for a patient.

```ruby
params = {
  :payer_id   => "000001",
  :provider_last_name => "Last",
  :provider_first_name => "First",
  :provider_npi => "12345678",
  :member_id => "12345678",
  :member_last_name => "Austen",
  :member_first_name => "Jane",
  :member_dob => "1955-12-14"
}
medicare = Eligible::Medicare.get(params)
medicare.to_hash # returns all coverage info for the request
medicare.error   # return error, if any
```

## Enrollment

Enrollment requests can have multiple enrollment NPIs. You can repeat
the enrollment for a NPI multiple times across different enrollment
requests.

### Reference
[https://reference.eligible.com/#enrollment-introduction](https://reference.eligible.com/#enrollment-introduction)

### Create an Enrollment Request

```ruby
params = {
  "service_provider_list" => [
      {
        "facility_name" => "Quality",
        "provider_name" => "Jane Austen",
               "tax_id" => "12345678",
              "address" => "125 Snow Shoe Road",
                 "city" => "Sacramento",
                "state" => "CA",
                  "zip" => "94107",
                 "ptan" => "54321",
                  "npi" => "987654321"
      },
      {
        "facility_name" => "Aetna",
        "provider_name" => "Jack Austen",
               "tax_id" => "12345678",
              "address" => "985 Snow Shoe Road",
                 "city" => "Menlo Park",
                "state" => "CA",
                  "zip" => "94107",
                 "ptan" => "54321",
                  "npi" => "987654321"
      }
    ],
    "payer_ids" => [
      "00431",
      "00282"
    ]
}
result = Eligible::Enrollment.post(params)
result.to_hash # returns the api call results
result.error   # return error, if any
```

### Retrieve an Enrollment Request

```ruby
params = { :enrollment_request_id => 123 }
enrollment = Eligible::Enrollment.get(params)
enrollment.to_hash # return the api call results
enrollment.error  # return error, if any
enrollment.enrollment_npis # quick access to the enrollment npis within the enrollment request object

params = { :npis => %w(123 456 789).join(',') }
enrollment = Eligible::Enrollment.get(params)
```

## Claims

### Reference

[https://reference.eligible.com/#create-a-claim](https://reference.eligible.com/#create-a-claim)

### Create Claim object

```ruby
params = {
    "receiver" => {
        "name" => "AETNA",
        "id" => "60054"
    },
    "billing_provider" => {
        "taxonomy_code" => "332B00000X",
        "practice_name" => "Jane Austen Practice",
        "npi" => "1922222222",
        "address" => {
            "street_line_1" => "419 Fulton",
            "street_line_2" => "",
            "city" => "San Francisco",
            "state" => "CA",
            "zip" => "94102"
        },
        "tin" => "43291023"
    },
    "subscriber" => {
        "last_name" => "Franklin",
        "first_name" => "Benjamin",
        "member_id" => "12312312",
        "group_id" => "455716",
        "group_name" => "",
        "dob" => "1734-05-04",
        "gender" => "M",
        "address" => {
            "street_line_1" => "435 Sugar Lane",
            "street_line_2" => "",
            "city" => "Sweet",
            "state" => "OH",
            "zip" => "436233127"
        }
    },
    "payer" => {
        "name" => "AETNA",
        "id" => "60054",
        "address" => {
            "street_line_1" => "Po Box 981106",
            "street_line_2" => "",
            "city" => "El Paso",
            "state" => "TX",
            "zip" => "799981222"
        }
    },
    "claim" => {
        "total_charge_amount" => "275",
        "claim_frequency" => "1",
        "patient_signature_on_file" => "Y",
        "provider_plan_participation" => "A",
        "direct_payment_authorized" => "Y",
        "release_of_information" => "I",
        "service_lines" => [
            {
                "line_number" => "1",
                "service_start" => "2013-03-07",
                "service_end" => "2013-03-07",
                "place_of_service" => "11",
                "charge_amount" => "275",
                "product_service" => "99213",
                "qualifier" => "HC",
                "diagnosis_1" => "32723"
            }
        ]
    }
}

result = Eligible::Claim.post(params)
enrollment.to_hash # return the api call results
enrollment.error  # return error, if any
```

### Retrieve all Claim objects/acknowledgments

```ruby
claims = Eligible::Claim.all # returns acknowledgment information for all claims that have been submitted with the API key
```

### Retrieve individual Claim object/acknowledgment

```ruby
params = {
  :reference_id => "12345"
}

claim = Eligible::Claim.get(params) # returns acknowledgment information on an individual claim identified by its reference_id
```

## Payment Status

### Reference

[https://reference.eligible.com/#payment-status](https://reference.eligible.com/#payment-status)

### Retrieve  Payment status

```ruby
params = { :payer_id => '00001',
           :provider_tax_id => '4373208923',
           :provider_npi => '1234567890',
           :provider_first_name => 'Thomas',
           :provider_first_name => 'Thomas',
           :member_id => '123',
           :member_first_name => 'Benjamin',
           :member_last_name => 'Franklin',
           :member_dob => '01-01-1980',
           :charge_amount => '100.00',
           :start_date => '2013-03-25',
           :end_date => '2013-03-25' }

result = Eligible::Payment.get(params)
result.to_hash   # return the api call results
result.error     # return error, if any
```

## X12

### X12 post

```ruby
params = { :x12 => "ISA*00*          *00*          *ZZ*SENDERID       *ZZ*ELIGIB         *130610*0409*^*00501*100000001*0*T*:~GS*HS*SENDERID*ELIGIB*20130610*0409*1*X*005010X279A1~ST*270*0001*005010X279A1~BHT*0022*13*137083739083716126837*20130610*0409~HL*1**20*1~NM1*PR*2*UnitedHealthCare*****PI*112~HL*2*1*21*1~NM1*1P*1*AUSTEN*JANE****XX*1222494919~HL*3*2*22*0~TRN*1*1*1453915417~NM1*IL*1*FRANKLIN*BENJAMIN****MI*23412342~DMG*D8*17371207~DTP*291*D8*20130610~EQ*30~SE*13*0001~GE*1*1~IEA*1*100000001~" }

result = Eligible::X12.post(params)
```

## Tickets

### Reference

[https://reference.eligible.com/#create-a-ticket](https://reference.eligible.com/#create-a-ticket)

### Create a ticket

```ruby
params = {:priority => 'normal',
          :title => 'TITLE',
          :notification_email => 'admin@eligible.com',
          :body => 'Your comment'}
result = Eligible::Ticket.create params
result.to_hash # return the api call results
enrollment.error  # return error, if any
```

### Get a ticket

```ruby
ticket = Eligible::Ticket.get(:id => 1)
ticket.to_hash # return the api call result
ticket.error   # return error, if any
```

### Update a ticket

```ruby
params = { :id => 1,
           :priority => 'normal',
           :title => 'TITLE',
           :notification_email => 'your_email@test.com',
           :body => 'Your comment'}
result = Eligible::Ticket.update(params)
result.to_hash # return the api call results
enrollment.error  # return error, if any
```

### Get comments for a ticket

```ruby
comments = Eligible::Ticket.get(:id => 1)
comments.to_hash # return the api call result
comments.error   # return error, if any

```

### Delete a ticket
```ruby
result = Eligible::Ticket.delete(:id => 1)
comments.to_hash # return the api call result
comments.error   # return error, if any
```

### Get all tickets

```ruby
Eligible::Ticket.all
```

## Customer

### Reference

[https://reference.eligible.com/#customers-introduction](https://reference.eligible.com/#customers-introduction)

### Create a customer

```ruby
customer_params = { customer: { name: "ABC company", 
                                site_name: "ABC site name"
                              }
                  }
customer_response = Eligible::Customer.post(customer_params)
customer_response.to_json
```

### Get a customer

```ruby
customer_params = { customer_id: "123" }
customer_response = Eligible::Customer.get(customer_params)
customer_response.to_json
```

### Update a customer

```ruby
customer_params = { customer_id: "123",
                    customer: { site_name: "XYZ site name" }
                  }
customer_response = Eligible::Customer.update(customer_params)
customer_response.to_json
```

### Get all customers

```ruby
customer_params = {}
customer_response = Eligible::Customer.all(customer_params)
customer_response.to_json
```

## Received Pdf

### Reference

[https://reference.eligible.com/#view-received-pdf](https://reference.eligible.com/#view-received-pdf)

### Get received pdf

```ruby
params = { enrollment_npi_id: '123' }
response = Eligible::ReceivedPdf.get(params)
response.to_hash
```

### Download received pdf
By default, it downloads to /tmp/received_pdf.pdf
```ruby
params = { enrollment_npi_id: '123', filename: 'file_path_where_to_download' }
Eligible::ReceivedPdf.download(params)
```

## Original Signature Pdf

### Reference

[https://reference.eligible.com/#create-original-signature-pdf](https://reference.eligible.com/#create-original-signature-pdf)

### Get original signature pdf

```ruby
params = { enrollment_npi_id: '123' }
response = Eligible::OriginalSignaturePdf.get(params)
response.to_hash
```

### Create original signature pdf

```ruby
params = { enrollment_npi_id: '123' }
params[:file] = File.new('path_to_file')
response = Eligible::OriginalSignaturePdf.post(params)
response.to_hash
```

### Update original signature pdf

```ruby
params = { enrollment_npi_id: '123' }
params[:file] = File.new('path_to_new_file')
response = Eligible::OriginalSignaturePdf.update(params)
response.to_hash
```

### Download original signature pdf
By default, it downloads to /tmp/original_signature_pdf.pdf
```ruby
params = { enrollment_npi_id: '123', filename: 'file_path_where_to_download' }
Eligible::OriginalSignaturePdf.download(params)
```

### Delete original signature pdf

```ruby
params = { enrollment_npi_id: '123' }
response = Eligible::OriginalSignaturePdf.delete(params)
response.to_hash
```

## Payer

### Reference

[https://reference.eligible.com/#introduction](https://reference.eligible.com/#introduction)

### List all the payers

```ruby
response = Eligible::Payer.list({})
response.collect { |payer| payer.to_hash }
```

### View a single payer

```ruby
params = { payer_id: '12345' }
response = Eligible::Payer.get(params)
response.to_hash
```

### Search options for a payer

```ruby
params = { payer_id: '12345' }
response = Eligible::Payer.search_options(params)
response.to_hash
``` 

### Search options for all payers

```ruby
response = Eligible::Payer.search_options({})
response.collect { |payer| payer.to_hash }
```

## Precertification

### Reference

[https://reference.eligible.com/#precertification](https://reference.eligible.com/#precertification)

### Inquiry

```ruby
params = { provider_npi: '1234567893',
           member_id: 'ABCDEF',
           member_dob: '2016-03-04'
         }
response = Eligible::Precert.inquiry(params)
response.to_hash
```

### Create

```ruby
params = { 'requester' => { 'information' => 'test' },
           'subscriber' => { 'last_name' => 'XYZ',
                             'first_name' => 'AVC',
                             'id' => '231213'
                           },
           'event' => { 'provider' => 'information' },
           'services' => [{ 'service' => 'test' }]
         }
response = Eligible::Precert.create(params)
response.to_hash
```

## Referral

### Inquiry

```ruby
params = { provider_npi: '1234567893',
           member_id: 'ABCDEF',
           member_dob: '2016-03-04'
         }
response = Eligible::Referral.inquiry(params)
response.to_hash
```

### Create

```ruby
params = { 'requester' => { 'information' => 'test' },
           'subscriber' => { 'last_name' => 'XYZ',
                             'first_name' => 'AVC',
                             'id' => '231213'
                           },
           'event' => { 'provider' => 'information' },
           'services' => [{ 'service' => 'test' }]
         }
response = Eligible::Referral.create(params)
response.to_hash
```

## Errors

This is the list of errors thrown from the eligible ruby gem.

1. Eligible::EligibleError - Base class for the customized errors raised from Eligible gem.
2. Eligible::APIError - Raised when there is some invalid response from the api call. Raised for error codes other than 400, 401 and 404.
3. Eligible::APIConnectionError - Raised when there is some network issue like socket error, not able to connect to Eligible etc.
4. Eligible::InvalidRequestError - Raised when error code is 400 or 404.
5. Eligible::AuthenticationError - Raised when authentication fails. Mostly due to wrong api key.
6. NotImplementedError - Raised when the functionality you are trying to use doesn't exist.
7. ArgumentError - Raised when all the required params are not provided.


## Contributing

Running `rake` will run the test suite along with rubocop as a basic
style assessment. If you are going to submit a pull request, please
verify that all tests pass and there are no rubocop errors. Please add
additional tests for any features or fixes provided.

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Run tests (see above)
5. Push to the branch (`git push origin my-new-feature`)
6. Create new Pull Request


ChangeLog
2017-02-23 Eligible  <support@eligible.com>

	* 2.8.1
	- Upgrades internal rest-client dependency to 2.0.1
	- Fixes mime-type dependency issue with newer Rails projects

2017-02-23  Eligible  <support@eligible.com>

	* 2.8.0
	- Adds support for session tokens

2016-12-14  Eligible  <support@eligible.com>

	* 2.7.0
	- Changed default content type to application/json

2016-11-18  Eligible  <support@eligible.com>

	* 2.6.3
	- Added a new certificate fingerprint

2016-08-31  Eligible  <support@eligible.com>
	* 2.6.2
	- New APIs added in testing mode, no public-facing changes

2016-05-10  Eligible  <support@eligible.com>
	* 2.6.1
	- Suppressing ssl_verify_callback return code warning
	- Added new endpoints: received_pdf, precert and referral
	- Refactored the code and fixed couple of bugs
	- Documentation updates for the endpoints

2016-02-23  Eligible  <support@eligible.com>

	* 2.6.0
	- Added new endpoints customer, original signature pdf and payer.
	- Added specs
	- Bumped api version to 1.5
	- Fixed few bugs

2016-01-07  Eligible  <support@eligible.com>

	* 2.5.0
	- Added certificate pinning
	- Added rubocop and fixed style/linting/metrics issues
	- Migrated from outdated test-unit/mocha to current rspec for test suite

0000-00-00  Eligible  <support@eligible.com>

	* 2.4
	- Refactoring Code
	- More test cases
	- Removed legacy endpoint for *plans*, *coverage* should be used instead.
	- Removed legacy endpoint for *services*, *coverage* should be used instead.
	- List of contributors and documentation updated.
	- Gemfile updated, dependencies updated as well.
	- Removed json gem in favor of multi_json
	- Fixed the code to let the users make x12 requests at anytime.
	- New endpoint for Tickets

0000-00-00  Eligible  <support@eligible.com>

	* 2.3
	- New endpoint for Batch

0000-00-00  Eligible  <support@eligible.com>

	* 2.2
	- New endpoint for x12 POST

0000-00-00  Eligible  <support@eligible.com>

	* 2.1
	- New endpoint for payment status

0000-00-00  Eligible  <support@eligible.com>

	* 2.0
	- Additional endpoints for claims, enrollments, and coverage

0000-00-00  Eligible  <support@eligible.com>

	* 1.1
	- Additional endpoints for service/general and service/list
	- Support x12 format in params

0000-00-00  Eligible  <support@eligible.com>

	* 1.0
	- Initial release



lib/eligible.rb
require 'cgi'
require 'set'
require 'rubygems'
require 'openssl'
require 'net/https'

require 'rest_client'
require 'multi_json'

require 'eligible/version'
require 'eligible/encryptor'
require 'eligible/util'
require 'eligible/json'
require 'eligible/eligible_object'
require 'eligible/api_resource'
require 'eligible/coverage_resource'
require 'eligible/demographic'
require 'eligible/claim'
require 'eligible/enrollment'
require 'eligible/coverage'
require 'eligible/payment'
require 'eligible/x12'
require 'eligible/medicare'
require 'eligible/ticket'
require 'eligible/customer'
require 'eligible/original_signature_pdf'
require 'eligible/received_pdf'
require 'eligible/payer'
require 'eligible/preauth_resource'
require 'eligible/precert'
require 'eligible/referral'
require 'eligible/public_key'
require 'eligible/lockbox'
require 'eligible/session_token'

# Errors
require 'eligible/errors/eligible_error'
require 'eligible/errors/api_connection_error'
require 'eligible/errors/authentication_error'
require 'eligible/errors/api_error'
require 'eligible/errors/invalid_request_error'

module Eligible
  @@api_key = nil
  @@test = false
  @@api_version = '1.5'
  @@api_base = "https://gds.eligibleapi.com/v#{@@api_version}"
  @@fingerprints = %w(79d62e8a9d59ae687372f8e71345c76d92527fac
                      4b2c6888ede79d0ee47339dc6fab5a6d0dc3cb0e
                      de4cdd0aae26df71290f0373af18e9ee7ecff18c)

  def self.api_url(url = '')
    @@api_base + url.to_s
  end

  def self.api_key
    @@api_key
  end

  def self.api_key=(api_key)
    @@api_key = api_key
  end

  def self.api_base
    @@api_base
  end

  def self.api_base=(api_base)
    @@api_base = api_base
  end

  def self.test
    @@test ? 'true' : 'false'
  end

  def self.test=(is_test)
    @@test = is_test
  end

  def self.api_version=(version)
    @@api_version = version
  end

  def self.api_version
    @@api_version
  end

  def self.fingerprints
    @@fingerprints
  end

  def self.add_fingerprint(digest)
    $stderr.puts 'The embedded certificate fingerprint was modified. This should only be done if instructed to by eligible support staff'
    @@fingerprints << digest
  end

  def self.direct_response?(params)
    params[:format].is_a?(String) && params[:format].downcase == 'x12'
  end

  def self.test_key?(params)
    Util.key?(params, :test)
  end

  def self.api_key?(params)
    Util.key?(params, :api_key)
  end

  def self.request(method, url, api_key, params = {}, headers = {})
    api_key ||= @@api_key
    test = self.test
    api_key = Util.value(params, :api_key) if api_key?(params)
    test = Util.value(params, :test) if test_key?(params)

    fail AuthenticationError, 'No API key provided. (HINT: set your API key using "Eligible.api_key = <API-KEY>".' unless api_key

    lang_version = "#{RUBY_VERSION} p#{RUBY_PATCHLEVEL} (#{RUBY_RELEASE_DATE})"
    debug_info = {
      bindings_version: Eligible::VERSION,
      lang: 'ruby',
      lang_version: lang_version,
      platform: RUBY_PLATFORM,
      publisher: 'eligible',
      uname: uname
    }

    # GET requests, parameters on the query string
    # POST requests, parameters as json in the body
    url = api_url(url)
    case method.to_s.downcase.to_sym
    when :get, :head, :delete
      url += "?api_key=#{api_key}"
      if params && params.count > 0
        query_string = Util.flatten_params(params).collect { |key, value| "#{key}=#{Util.url_encode(value)}" }.join('&')
        url += "&#{query_string}"
      end
      url += "&test=#{test}"
      payload = nil
    else
      params.merge!('api_key' => api_key, 'test' => test)
      payload = Util.key?(params, :file) ? params : Eligible::JSON.dump(params)
    end

    begin
      headers = { x_eligible_debuginfo: Eligible::JSON.dump(debug_info) }.merge(headers)
    rescue => e
      headers = {
        x_eligible_client_raw_user_agent: debug_info.inspect,
        error: "#{e} (#{e.class})"
      }.merge(headers)
    end

    headers = {
      user_agent: "eligible-ruby/#{Eligible::VERSION}",
      authorization: "Bearer #{api_key}",
      content_type: 'application/json'
    }.merge(headers)

    headers[:eligible_version] = api_version if api_version

    opts = {
      method: method,
      url: url,
      headers: headers,
      open_timeout: 30,
      payload: payload,
      timeout: 80,
      ssl_verify_callback: verify_certificate,
      ssl_verify_callback_warnings: false
    }

    begin
      response = execute_request(opts)

    rescue SocketError => e
      handle_restclient_error(e)

    rescue NoMethodError => e
      # Work around RestClient bug
      raise unless e.message =~ /\WRequestFailed\W/

      e = APIConnectionError.new('Unexpected HTTP response code')
      handle_restclient_error(e)

    rescue RestClient::ExceptionWithResponse => e
      err_rcode = e.http_code
      err_rbody = e.http_body

      if err_rcode && err_rbody
        handle_api_error(err_rcode, err_rbody)
      else
        handle_restclient_error(e)
      end

    rescue RestClient::Exception, Errno::ECONNREFUSED => e
      handle_restclient_error(e)
    end

    rbody = response.body
    rcode = response.code

    begin
      # Would use :symbolize_names => true, but apparently there is
      # some library out there that makes symbolize_names not work.
      resp = direct_response?(params) ? rbody : Eligible::JSON.load(rbody)
    rescue MultiJson::DecodeError
      raise APIError.new("Invalid response object from API: #{rbody.inspect} (HTTP response code was #{rcode})", rcode, rbody)
    end

    resp = Util.symbolize_names(resp)
    return [ resp, api_key ]
  end

  def self.verify_certificate
    lambda do |preverify_ok, certificate_store|
      return true if test == 'true'
      return false unless preverify_ok
      received = certificate_store.chain.first
      return true unless received.to_der == certificate_store.current_cert.to_der
      valid_fingerprint?(received)
    end
  end

  def self.valid_fingerprint?(received)
    fingerprints.include?(OpenSSL::Digest::SHA1.hexdigest(received.to_der))
  end

  private

  def self.uname
    @@uname ||= RUBY_PLATFORM =~ /linux|darwin/i ? `uname -a 2>/dev/null`.strip : nil
  end

  def self.execute_request(opts)
    RestClient::Request.execute(opts)
  end

  def self.error_message(error)
    return error.to_s unless error.is_a?(Hash)
    result = error[:details] || error[:reject_reason_description] || error
    return result.to_s
  end

  def self.handle_api_error(rcode, rbody)
    begin
      error_obj = Eligible::JSON.load(rbody)
      error_obj = Util.symbolize_names(error_obj)
      fail EligibleError unless error_obj.key?(:error)
      error = error_obj[:error]
    rescue MultiJson::DecodeError, EligibleError
      raise APIError.new("Invalid response object from API: #{rbody.inspect} (HTTP response code was #{rcode})", rcode, rbody)
    end

    error_msg = error_message(error)

    case rcode
    when 400, 404 then
      raise invalid_request_error(error_msg, rcode, rbody, error_obj)
    when 401
      raise authentication_error(error_msg, rcode, rbody, error_obj)
    else
      raise api_error(error_msg, rcode, rbody, error_obj)
    end
  end

  def self.invalid_request_error(error_msg, rcode, rbody, error_obj)
    InvalidRequestError.new(error_msg, rcode, rbody, error_obj)
  end

  def self.authentication_error(error_msg, rcode, rbody, error_obj)
    AuthenticationError.new(error_msg, rcode, rbody, error_obj)
  end

  def self.api_error(error_msg, rcode, rbody, error_obj)
    APIError.new(error_msg, rcode, rbody, error_obj)
  end

  def self.handle_restclient_error(e)
    case e
    when RestClient::ServerBrokeConnection, RestClient::RequestTimeout
      message = "Could not connect to Eligible (#{@@api_base}).  Please check your internet connection and try again."
    when RestClient::SSLCertificateNotVerified
      message = "Could not verify Eligible's SSL certificate."
    when SocketError
      message = 'Unexpected error communicating when trying to connect to Eligible.'
    else
      message = 'Unexpected error communicating with Eligible. If this problem persists, let us know at support@eligible.com.'
    end
    fail APIConnectionError, "#{message}\n\n(Network error: #{e.message})"
  end
end
